### PR TITLE
Raise 501 for Unauthenticated Artifact Stores

### PR DESCRIPTION
**Pull request by fa9r**
### Main Change
**artifact_utils.py Lines 325-332:** Bugfix following #1472. If the artifact store holding the visualization does not have authentication credentials built in, it would previously raise an implementation-specific 500 error, but we want to raise 501 so we can catch it in the frontend.

### Minor Changes
While at it, I also refactored the code a bit:
- renamed `materializer_utils.py` > `artifact_utils.py`
- split artifact visualization logic into two new helper functions `_load_artifact_store_of_artifact` and `_load_uri_from_artifact_store` that can be reused to load other things from the artifact store in the future.
- Moved type checking imports where they belong

## Pre-requisites
Please ensure you have done the following:
- [x] I have read the **CONTRIBUTING.md** document.
- [ ] If my change requires a change to docs, I have updated the documentation accordingly.
- [ ] If I have added an integration, I have updated the [integrations](https://docs.zenml.io/component-gallery/integrations) table and the [corresponding website section](https://zenml.io/integrations).
- [ ] I have added tests to cover my changes.

## Types of changes

- [x] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] Other (add details above)

### CODE REVIEW

File: src/zenml/utils/artifact_utils.py [hunk 1]
```diff
@@ -243,30 +280,53 @@ def load_artifact_visualization(
...
     except FileNotFoundError:
         raise DoesNotExistException(
-            f"Artifact '{artifact.id}' could not be visualized because the "
-            f"visualization file '{visualization.uri}' was not be found in "
-            f"the underlying artifact store '{artifact_store_model.name}'."
+            f"File '{uri}' does not exist in artifact store "
+            f"'{artifact_store.name}'."
+        )
+    except Exception:
+        link = "https://docs.zenml.io/component-gallery/artifact-stores/custom#enabling-artifact-visualizations-with-custom-artifact-stores"
```
Comment (stefannica) [inline src/zenml/utils/artifact_utils.py:326]: Could you also please do a logger.exception() here so we don't lose the error info ? catching generic exceptions and then ignoring their source can be devastating for maintainability.

**Review by stefannica — CHANGES_REQUESTED** [review]

File: src/zenml/utils/artifact_utils.py [hunk 2]
```diff
@@ -215,24 +216,60 @@ def load_artifact_visualization(
...
+
+
+def _load_artifact_store_of_artifact(
+    artifact: "ArtifactResponseModel",
+    zen_store: Optional["BaseZenStore"] = None,
+) -> "BaseArtifactStore":
+    """Load the artifact store of the given artifact.
+
+    Args:
+        artifact: The artifact to load the artifact store of.
```
Comment (strickvl) [inline src/zenml/utils/artifact_utils.py:257]:
```suggestion
        artifact: The artifact for which to load the artifact store.
```

;)

File: src/zenml/utils/artifact_utils.py [hunk 3]
```diff
@@ -215,24 +216,60 @@ def load_artifact_visualization(
...
+        zen_store: The ZenStore to use for finding the artifact store. If not
+            provided, the ZenStore of the client will be used.
```
Comment (strickvl) [inline src/zenml/utils/artifact_utils.py:259]:
```suggestion
            provided, the client's ZenStore will be used.
```

File: src/zenml/utils/artifact_utils.py [hunk 4]
```diff
@@ -243,30 +280,53 @@ def load_artifact_visualization(
...
+    uri: str,
+    artifact_store: "BaseArtifactStore",
+    mode: str = "rb",
+) -> Any:
+    """Load the given uri from the given artifact store.
+
+    Args:
+        uri: The uri of the file to load.
+        artifact_store: The artifact store to load the file from.
+        mode: The mode to open the file in.
```
Comment (strickvl) [inline src/zenml/utils/artifact_utils.py:308]:
```suggestion
        artifact_store: The artifact store from which to load the file.
        mode: The mode in which to open the file.
```

File: src/zenml/utils/artifact_utils.py [hunk 5]
```diff
@@ -243,30 +280,53 @@ def load_artifact_visualization(
...
+
+    Returns:
+        The loaded file.
```
Comment (strickvl) [inline src/zenml/utils/artifact_utils.py:311]: are we returning a loaded file, or a URI?

**Review by strickvl — CHANGES_REQUESTED** [review]

File: src/zenml/utils/artifact_utils.py (hunk 1 above)
Comment (fa9r) [reply src/zenml/utils/artifact_utils.py:327]: Done 👍

File: src/zenml/utils/artifact_utils.py (hunk 5 above)
Comment (fa9r) [reply src/zenml/utils/artifact_utils.py:311]: A file. I guess the function name was confusing, I adjusted it accordingly.

**Review by strickvl — APPROVED** [review]
🚀

**Review by htahir1 — APPROVED** [review]

**Pull request merged**